packages/kit/types/index.d.ts

```diff
@@ -422,6 +422,19 @@ export interface KitConfig {
 	 * @default ".svelte-kit"
 	 */
 	outDir?: string;
+	/**
+	 * Options related to the build output format
+	 */
+	output?: {
+		/**
+		 * What preload strategy to use for JavaScript files to avoid waterfalls:
+		 * - `modulepreload` - uses `<link rel="modulepreload">` to preload JavaScript files. Works only in Chromium-based browsers currently, and soon in Safari, too.
+		 * - `preload-js` - uses `<link rel="preload">` to preload JavaScript files. Works in all browsers but Firefox, and causes double-parsing of the script for Chromium-based browser.
+		 * - `preload-mjs` - uses `<link rel="preload">` to preload JavaScript files, but uses the `.mjs` extension. Works in Chromium-based browsers and Safari. Check of your provider/CDN has the correct MIME type for `.mjs` files before turning this on.
+		 * @default "modulepreload"
+		 */
```

Comment:
I think we can afford to add loads of context about these different options and what they entail

```suggestion
		/**
		 * SvelteKit will preload the JavaScript modules needed for the initial page to avoid import 'waterfalls', resulting in faster application startup. There
		 * are three strategies with different trade-offs:
		 * - `modulepreload` - uses `<link rel="modulepreload">`. This delivers the best results in Chromium-based browsers, but is currently ignored by Firefox and Safari (though support is coming to Safari soon).
		 * - `preload-js` - uses `<link rel="preload">`. Prevents waterfalls in Chromium and Safari, but Chromium will parse each module twice (once as a script, once as a module). Causes modules to be requested twice in Firefox. This is a good setting if you want to maximise performance for users on iOS devices at the cost of a very slight degradation for Chromium users.
		 * - `preload-mjs` - uses `<link rel="preload">` but with the `.mjs` extension which prevents double-parsing in Chromium. Some static webservers will fail to serve .mjs files with the correct MIME type, which will cause your application to break. If that doesn't apply to you, this is the option that will deliver the best performance for the largest number of users, until `modulepreload` is more widely supported.
		 * @default "modulepreload"
		 */
```

Reply:
Forgive me for joining this discussion late, but is there something I can read that explains the problem and the various solutions, and what you’re trying to achieve? Like are you trying to get the browser to download these other JavaScript files and parse them but not evaluate them?

For example, it occurs to me that you could just add a `script` block in the output HTML, like

```html
<script type="module">
import '_app/start.js' // Or whatever the generated full or root-prefixed path is for each chunk
</script>
```

But this would not only download but also evaluate each file (and its dependencies), which I assume you’re trying to avoid? You could also just `fetch` each one, to download and cache but not evaluate (though does a later `import` redownload?) or use dynamic `import()` to apply custom logic to determining which chunks to download and evaluate.

Basically I’m just wondering what your goal is and the various alternatives you tried, and what you found when you investigated each one. If that were written down somewhere (and maybe it already is), it would help others to try to improve on what you’ve implemented here. It would also help if there were a test in your suite to determine if what you’re trying to avoid, the “waterfalls,” happens or not.

Reply:
The documentation in the comment above _is_ the explanation. We need to retain control of evaluation order because we can't run user code until some setup work has taken place — we _could_ refactor the code instead but it would increase bundle sizes, and could potentially box us in in future. `modulepreload` is the standard way to prevent import waterfalls.

`fetch`-based solutions don't work, unfortunately — things go into a different 'bucket' depending on whether they were fetched or imported.